explorer/interpreter/type_checker.cpp

```diff
@@ -436,40 +437,68 @@ auto TypeChecker::FieldTypesImplicitlyConvertible(
-  while (curr_class_type->base().has_value()) {
-    curr_class_type = curr_class_type->base().value();
-    all_classes.push_back(curr_class_type);
-  }
-  return all_classes;
-}
-
 auto TypeChecker::FieldTypes(const NominalClassType& class_type) const
     -> std::vector<NamedValue> {
   std::vector<NamedValue> field_types;
-  for (const auto class_type : GetClassHierarchy(class_type)) {
-    for (Nonnull<Declaration*> m : class_type->declaration().members()) {
-      switch (m->kind()) {
-        case DeclarationKind::VariableDeclaration: {
-          const auto& var = cast<VariableDeclaration>(*m);
-          Nonnull<const Value*> field_type =
-              Substitute(class_type->bindings(), &var.binding().static_type());
-          field_types.push_back(
-              {.name = var.binding().name(), .value = field_type});
-          break;
-        }
-        default:
-          break;
+  for (Nonnull<Declaration*> m : class_type.declaration().members()) {
+    switch (m->kind()) {
+      case DeclarationKind::VariableDeclaration: {
+        const auto& var = cast<VariableDeclaration>(*m);
+        Nonnull<const Value*> field_type =
+            Substitute(class_type.bindings(), &var.binding().static_type());
+        field_types.push_back(
+            {.name = var.binding().name(), .value = field_type});
+        break;
       }
+      default:
+        break;
     }
   }
   return field_types;
 }
 
+auto TypeChecker::FieldTypesWithBase(const NominalClassType& class_type) const
+    -> std::vector<NamedValue> {
+  auto fields = FieldTypes(class_type);
+  if (class_type.base().has_value()) {
+    auto base_fields = FieldTypesWithBase(*class_type.base().value());
+    fields.emplace_back(
+        NamedValue{.name = NominalClassValue::base_field,
+                   .value = (new StructType(std::move(base_fields)))});
+  }
+  return fields;
+}
+
+auto TypeChecker::StructImplicitlyConvertibleToClass(
+    const StructType& source_struct, const NominalClassType& dest_class,
+    const ImplScope& impl_scope, bool allow_user_defined_conversions) const
+    -> bool {
+  std::vector<NamedValue> struct_fields{source_struct.fields()};
+  const auto base_it = std::find_if(
+      struct_fields.begin(), struct_fields.end(), [](const auto& field) {
+        return field.name == NominalClassValue::base_field;
+      });
+  std::optional<const Value*> base_value;
+  if (base_it != struct_fields.end()) {
+    base_value = base_it->value;
+    struct_fields.erase(base_it);
+  }
+  if (base_value.has_value() != dest_class.base().has_value()) {
+    return false;
+  }
+  if (!FieldTypesImplicitlyConvertible(struct_fields, FieldTypes(dest_class),
+                                       impl_scope)) {
+    return false;
+  }
+  if (base_value.has_value()) {
+    if (!IsImplicitlyConvertible(base_value.value(), dest_class.base().value(),
+                                 impl_scope, allow_user_defined_conversions)) {
+      return false;
+    }
+  }
```

Comment:
```suggestion
  if (base_value.has_value() && !IsImplicitlyConvertible(base_value.value(), dest_class.base().value(),
                                 impl_scope, allow_user_defined_conversions)) {
      return false;
  }
```

I won't comment on the `if` series, but I'll always prefer && instead of nesting.

Reply:
Yeah its not the most elegant code. Its grouped logically, putting all the conditions together makes it very hard to read or make sense of. But definitely agree for this one.